File: code/go/02-structured-output/main.go
// Entry point for code/go/02-structured-output.
//
// Use -mode to select which example to run:
//
//	-mode=template           Prompt template demo
//	-mode=few-shot          Zero-shot vs few-shot comparison
//	-mode=chain-of-thought  Reasoning with and without CoT
//	                        (docs/01-foundations/02-prompt-engineering.md)
//	-mode=structured        Structured output: schema-validated extraction
//	                        (docs/01-foundations/03-structured-output.md)
//
// Default is -mode=template.
package main

import (
	"flag"
	"fmt"
)

func main() {
	mode := flag.String("mode", "template", "Mode: template, few-shot, chain-of-thought, or structured")
	flag.Parse()

	switch *mode {
	case "template":
		runPromptTemplate()
	case "few-shot":
		RunFewShotComparison()
	case "chain-of-thought":
		RunChainOfThought()
	case "structured":
		runStructuredExtraction()
	default:
		fmt.Printf("Unknown mode %q. Use -mode=template, -mode=few-shot, -mode=chain-of-thought, or -mode=structured\n", *mode)
	}
}

File: code/python/02-structured-output/main.py
"""Entry point for the chapter-02 prompt-engineering demos.

Use --mode to switch between the three examples referenced by the chapter.
Structured-output examples still run directly from their own scripts.
"""

from __future__ import annotations

import argparse

from chain_of_thought import main as chain_of_thought_main
from few_shot_comparison import main as few_shot_main
from prompt_template import main as prompt_template_main


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--mode",
        choices=("template", "few-shot", "chain-of-thought"),
        default="template",
        help="Which chapter-02 example to run.",
    )
    return parser.parse_args()

if __name__ == "__main__":
    args = parse_args()
    if args.mode == "template":
        prompt_template_main()
    elif args.mode == "few-shot":
        few_shot_main()
    else:
        chain_of_thought_main()
